### PR TITLE
Required prop fix

### PR DESCRIPTION
**Pull request by Fluidbyte**
Fixes issue with `require`'d props by:

- Default `required` state `false`
- Only runs full props checks on values that exist (not `undefined`) and are not required

Closes #26

### CODE REVIEW

**Comment by TomFrost** [comment]
Since `modifier` runs _after_ `generator` and `default` in the normal flow, do you think it's worth including in the noVal flow as well? It seems silly, but if modifiers will run on default values in one case, it might be counterintuitive if they don't in the second case.

<hyperpedantic>The noVal flow is also redefining objects that have already been defined once in the normal flow; could be ever so slightly DRYer if those objects were defined out of scope and then included by reference. But this is insane nit-picky so I understand if the current way is better for readability</hyperpedantic>

**Comment by Fluidbyte** [comment]
On your first point - I thought about this _a lot_ and my conclusion was "no". If you run a modifier on a value that came from a generator, then your generator is not solid. It wouldn't be hard to add this, but I'd like to see a demand for it.

As for the second point, I agree, I'll clean that up a bit and make it DRY'er.

**Comment by TomFrost** [comment]
Ha, my second point was surrounded in `<hyperpedantic></hyperpedantic>` tags, but it looks like github interpreted them as actual html.

To play devil's advocate for the first point, it might be pretty easy for someone to unknowingly have the modifier doing something to a default value (One engineer sets lastName default to "doe". Another engineer picks up ticket a month later to auto-capitalize the first letter of every last name with a modifier. `doe` becomes `Doe`, everyone's happy. Until someone realizes that, since a default is set, `required: true` isn't necessary anymore, and they remove it, and now they're getting lowercased `doe` again for no obvious reason. (This is an extreme case, so I'm open to counterpoint-- I just tend to err on the side of eliminating even far-flung gotchas)

One other item I just noticed: we just silently made it the Type's responsibility to make sure the value isn't undefined; that means that `required: true` has to be enforced once for every type. Chances are that would happen naturally anyway, but before we had an implicit contract that undefines would be caught before that point. Not sure if that's worth considering.

**Comment by Fluidbyte** [comment]
Alright, I'll add modifiers to the run when not required and undefined. It does make results more "expected".

On the second point - yeah... I went back and forth. I feel like opt-in required's are much more explicit and more of a standard practice in other validation libs. My thought is to address this in another issue; create a way (similar to, or piggybacking off `strict` mode) to allow model-wide `require`'d to be set.

**Comment by TomFrost** [comment]
Don't we already have a solution for that? It was there, but got deleted.

**Comment by Fluidbyte** [comment]
Not that I can think of. Strict enforces that a property must be defined, but we don't have anything for defaulting required.

**Comment by TomFrost** [comment]
OHohohoh we're talking about two different things. I agree that we wouldn't want to make `require: true` the default (I don't think it was before, even). What I'm talking about is that _enforcing_ `required: true` used to happen as a prop, but that prop got deleted. The only place it gets caught, now, is in the type check assuming that the type tests for it.

File: src/lib/validators.js [hunk 1]
```diff
@@ -59,18 +59,6 @@ const validators = {
...
-
-  /**
-   * Validator requires method, used by model
-   * @param {Object} def The property configuration
-   * @param {String} key The key name of the property
-   * @param {*} value The value being validated
-   */
-  required: function(def, key, value) {
-    if (def.required && !value) {
-      this.errors.push({ type: 'required', sub: 'default', key, value, message: `Property '${key}' is required` })
```
Comment (TomFrost) [inline src/lib/validators.js]: To illustrate my last comment, this was deleted and never re-implemented.

**Comment by Fluidbyte** [comment]
Right, it doesn't need it - unless the type allows `undefined` explicitly it's going to fail; regex's won't `match`, and conditionals won't pass, etc. The only exception to this is the `any` type which just passes through the value.

**Comment by Fluidbyte** [comment]
I removed it because in all the tests I created the `required` validator was just creating a useless error since the type test had already failed.

**Comment by TomFrost** [comment]
Ahh, I thought we were running `required` _before_ the type, that makes sense. I brought it up because I thought "[keyname] must be of type 'string'" was a weird error to see when the actual issue was "Hey, you didn't define [keyname] and you need to"

**Pull request merged**